mlflow/utils/async_logging/async_logging_queue.py

```diff
@@ -127,27 +134,26 @@ def _fetch_batch_from_queue(self) -> List[RunBatch]:
         batches = []
         if self._queue.empty():
             return batches
-        queue_size = self._queue.qsize()  # Estimate the queue's size.
         merged_batch = self._queue.get()
-        for i in range(queue_size - 1):
-            if self._queue.empty():
-                # `queue_size` is an estimate, so we need to check if the queue is empty.
-                break
-            batch = self._queue.get()
-            if (
-                merged_batch.run_id != batch.run_id
-                or len(merged_batch.metrics) + len(batch.metrics) >= 1000
-                or len(merged_batch.params) + len(batch.params) >= 100
-                or len(merged_batch.tags) + len(batch.tags) >= 100
-            ):
-                # Make a new batch if the run_id is different or the batch is full.
-                batches.append(merged_batch)
-                merged_batch = batch
-            else:
-                merged_batch.add_child_batch(batch)
-                merged_batch.params.extend(batch.params)
-                merged_batch.tags.extend(batch.tags)
-                merged_batch.metrics.extend(batch.metrics)
+        # avoid flakiness when fetching batch from queue
```

Comment:
can you elaborate on this?

Reply:
from my test, this hangs (and it's not always hanging). From the documentation both self._queue.qsize and self._queue.empty() is not trustable, so I retry 3 times here to make sure the queue is actually empty.

Reply:
which line hangs?

Reply:
I actually didn't go deeper, the test `test_tf_keras_autolog_implicit_batch_size_works` hangs, and I assume `log_batch_async` is the cause. I updated this and seems the test can consistently pass

Reply:
Instead of relying on `empty`, can we use `Queue.join`?

https://docs.python.org/3/library/queue.html#queue.Queue.join

Reply:
Had another try, seems this is not relevant (can't repro the hang even without this change). But guess we can't use join because this is used for async_logging_buffer_seconds